x-pack/plugins/enterprise_search/public/applications/workplace_search/views/content_sources/components/add_source/add_source.test.tsx

```diff
@@ -0,0 +1,164 @@
+
+jest.mock('../../../../../shared/kibana', () => ({
+  KibanaLogic: { values: { navigateToUrl: jest.fn() } },
+}));
+import { KibanaLogic } from '../../../../../shared/kibana';
+
+import React from 'react';
+import { shallow } from 'enzyme';
+
+import { Loading } from '../../../../../shared/loading';
+
+import { AddSource } from './add_source';
+import { AddSourceSteps } from './add_source_logic';
+import { ConfigCompleted } from './config_completed';
+import { ConfigurationIntro } from './configuration_intro';
+import { ConfigureCustom } from './configure_custom';
+import { ConfigureOauth } from './configure_oauth';
+import { ConnectInstance } from './connect_instance';
+import { ReAuthenticate } from './re_authenticate';
+import { SaveConfig } from './save_config';
+import { SaveCustom } from './save_custom';
+
+describe('AddSourceList', () => {
+  const initializeAddSource = jest.fn();
+  const setAddSourceStep = jest.fn();
+  const saveSourceConfig = jest.fn((_, setConfigCompletedStep) => {
+    setConfigCompletedStep();
+  });
+  const createContentSource = jest.fn((_, formSubmitSuccess) => {
+    formSubmitSuccess();
+  });
+  const resetSourcesState = jest.fn();
+
+  const mockValues = {
+    addSourceCurrentStep: AddSourceSteps.ConfigIntroStep,
+    sourceConfigData,
+    dataLoading: false,
+    newCustomSource: {},
+    isOrganization: true,
+  };
+
+  beforeEach(() => {
+    setMockActions({
+      initializeAddSource,
+      setAddSourceStep,
+      saveSourceConfig,
+      createContentSource,
+      resetSourcesState,
+    });
+    setMockValues(mockValues);
+  });
+
+  it('renders default state', () => {
+    const wrapper = shallow(<AddSource sourceIndex={1} />);
+    wrapper.find(ConfigurationIntro).prop('advanceStep')();
+
+    expect(setAddSourceStep).toHaveBeenCalledWith(AddSourceSteps.SaveConfigStep);
+  });
+
+  it('handles loading state', () => {
+    setMockValues({ ...mockValues, dataLoading: true });
+    const wrapper = shallow(<AddSource sourceIndex={1} />);
+
+    expect(wrapper.find(Loading)).toHaveLength(1);
+  });
+
+  it('renders Config Completed step', () => {
+    setMockValues({
+      ...mockValues,
+      addSourceCurrentStep: AddSourceSteps.ConfigCompletedStep,
+    });
+    const wrapper = shallow(<AddSource sourceIndex={1} />);
+    wrapper.find(ConfigCompleted).prop('advanceStep')();
+
+    expect(KibanaLogic.values.navigateToUrl).toHaveBeenCalledWith(
+      '/sources/add/confluence_cloud/connect'
+    );
+    expect(setAddSourceStep).toHaveBeenCalledWith(AddSourceSteps.ConnectInstanceStep);
+  });
+
+  it('renders Save Config step', () => {
+    setMockValues({
+      ...mockValues,
+      addSourceCurrentStep: AddSourceSteps.SaveConfigStep,
+    });
+    const wrapper = shallow(<AddSource sourceIndex={1} />);
+    const saveConfig = wrapper.find(SaveConfig);
+    saveConfig.prop('advanceStep')();
+    saveConfig.prop('goBackStep')!();
```

Comment:
Is `!` here a non-null assertion operator? https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

Reply:
That's correct. I got this TypeScript error without it:
![image](https://user-images.githubusercontent.com/1869731/102650245-34806f00-4130-11eb-942a-1fbe9dc8abb0.png)

Reply:
Thanks! Looks a bit weird without a dot, I have to say 😄